crates/sidevm/host-runtime/src/env.rs

```diff
@@ -620,24 +665,27 @@ fn sidevm_ocall_fast_return(
 
 // Support all ocalls. Put the result into a temporary vec and wait for next fetch_result ocall to fetch the result.
 fn sidevm_ocall(
-    env: &Env,
+    mut func_env: FunctionEnvMut<Env>,
     task_id: i32,
     func_id: i32,
     p0: IntPtr,
     p1: IntPtr,
     p2: IntPtr,
     p3: IntPtr,
 ) -> Result<IntRet, OcallAborted> {
-    let mut env = env.inner.lock().unwrap();
-    let env = &mut *env;
-
-    env.state.current_task = task_id;
-    let result = set_task_env(env.state.awake_tasks.clone(), task_id, || {
-        env::dispatch_call(&mut env.state, &env.memory, func_id, p0, p1, p2, p3)
+    let inner = func_env.data().inner.clone();
+    let mut guard = inner.lock().unwrap();
+    let env = &mut *guard;
+
+    env.current_task = task_id;
+    let result = set_task_env(env.awake_tasks.clone(), task_id, || {
+        let mut state = env.make_mut(&mut func_env);
+        env::dispatch_call(&mut state, func_id, p0, p1, p2, p3)
     });
-    if env.state.ocall_trace_enabled {
+    let env = &mut *guard;
```

Comment:
Why to get the mut reference again here?

Reply:
The first borrow was moved into the closure, so we need to borrow again. The `mut` in the second can be removed though.

Reply:
Removed.

Strange, it's happy to compile now even the second borrow was removed.
The compiler told me the `env` was moved into the closure yesterday. Maybe it was confused by other errors at that time.

Reply:
Nice